### PR TITLE
[GEOS-10724] SpringResourceAdaptor should throw a FileNotFoundException instead of creating any missing file

### PR DESCRIPTION
**Pull request by dieterstueken**
[GEOS-10724](https://osgeo-org.atlassian.net/browse/GEOS-10724)

The `SpringResourceAdaptor` wraps an `org.geoserver.platform.resource.Resource` into a `org.springframework.core.io.Resource`.

Both are quite similar, but while the geoserver Resource wraps any `IOException` into an `IllegalStateException`  
(which is fine to use it in λ-expressions), the spring Resource however can and should throw the `IOException` itself.

Looking at [GeoServerPropertyConfigurer:104](https://github.com/geoserver/geoserver/blob/190f8849b9a2da54f709f227f3ad3e93a1194f41/src/main/src/main/java/org/geoserver/config/GeoServerPropertyConfigurer.java#L104) it seems, the missing config file shall be created and filled.
However, `super.loadProperties` indirectly creates some empty config file instead. 

Since this situation was not covered before, the first commit adds some unit test to show up the problem.

The second commit modifies `SpringResourceAdaptor`  to throw `FileNotFoundException` for `Resource.Type.UNDEFINED`.

Remaining question: what about `Resource.Type.DIRECTORY`?

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary)
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/config/SpringResourceAdaptor.java [hunk 1]
```diff
@@ -74,22 +86,30 @@ public boolean isOpen() {
...
     @Override
     public URL getURL() throws IOException {
-        return Resources.find(resource).toURI().toURL();
+        return getFile().toURI().toURL();
     }
 
     @Override
     public URI getURI() throws IOException {
-        return Resources.find(resource).toURI();
+        return getFile().toURI();
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/config/SpringResourceAdaptor.java:94]: Okay I see what you did there; using getFile() to do the null check in a single location

File: src/main/src/main/java/org/geoserver/config/SpringResourceAdaptor.java [hunk 2]
```diff
@@ -74,22 +86,30 @@ public boolean isOpen() {
...
+        if (file == null) {
+            throw new FileNotFoundException(resource.path());
+        }
+        return file;
     }
 
     @Override
     public long contentLength() throws IOException {
-        return Resources.find(resource).length();
+        return getFile().length();
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/config/SpringResourceAdaptor.java:112]: It is these kind of method where I want to check the javadocs; and see if we could return ``0`` or ``-1`` or something to be kinder to client code. Still this is a judgement call and you are the person doing the work so I trust your judgement :)

File: src/main/src/test/java/org/geoserver/config/SpringResourceAdaptorTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
+/* (c) 2015 Open Source Geospatial Foundation - all rights reserved
```
Comment (jodygarnett) [inline src/main/src/test/java/org/geoserver/config/SpringResourceAdaptorTest.java:1]: Probably ``(c) 2022`` ?

**Review by jodygarnett — COMMENTED** [review]
I like the direction this is heading

File: src/main/src/main/java/org/geoserver/config/SpringResourceAdaptor.java (hunk 2 above)
Comment (dieterstueken) [reply src/main/src/main/java/org/geoserver/config/SpringResourceAdaptor.java:112]: yes, I see. I was too fast to replace them all. I will check again.

**Comment by dieterstueken** [comment]
I was unable to figure out the reason for those  _PMD violations_. How can I reproduce it locally?
Also, the repeated  _bean lookup error_ occurs randomly. 
Did not find the cause and it seems not to be related to my changes.

**Comment by aaime** [comment]
Regarding PMD, see https://docs.geoserver.org/latest/en/developer/qa-guide/index.html (linked from the checklist too).

File: src/main/src/test/java/org/geoserver/config/GeoServerPropertyConfigurerTest.java [hunk 4]
```diff
@@ -35,6 +37,17 @@ public void tearDown() throws Exception {
         ctx.close();
     }
 
+    @Test
+    public void testConfigFileCreated() throws IOException {
```
Comment (sikeoka) [inline src/main/src/test/java/org/geoserver/config/GeoServerPropertyConfigurerTest.java:41]: Is this testing anything relevant to the rest of the PR?

**Comment by sikeoka** [comment]
> I was unable to figure out the reason for those _PMD violations_. How can I reproduce it locally? Also, the repeated _bean lookup error_ occurs randomly. Did not find the cause and it seems not to be related to my changes.

PMD Failure: org.geoserver.config.SpringResourceAdaptorTest:8 Rule:wildcards Priority:3 No Wildcard Imports.

**Comment by jodygarnett** [comment]
> > I was unable to figure out the reason for those _PMD violations_. How can I reproduce it locally?
> 
> PMD Failure: org.geoserver.config.SpringResourceAdaptorTest:8 Rule:wildcards Priority:3 No Wildcard Imports. PMD Failure: org.geoserver.config.SpringResourceAdaptorTest:9 Rule:wildcards Priority:3 No Wildcard Imports.

Sorry I should of seen this earlier - I struggled with this also and added some examples on how to run PMD to the user guide: https://docs.geoserver.org/latest/en/developer/qa-guide/index.html


```
mvn verify -Ppmd
```

or
```
mvn initialize pmd:check -Ppmd
```

File: src/main/src/test/java/org/geoserver/config/GeoServerPropertyConfigurerTest.java (hunk 4 above)
Comment (jodygarnett) [reply src/main/src/test/java/org/geoserver/config/GeoServerPropertyConfigurerTest.java:41]: Did we decided if this is needed here?

**Review by jodygarnett — APPROVED** [review]
There is one question about GeoServerPropertyConfigurerTest, but the change here seems good and has test coverage.

File: src/main/src/test/java/org/geoserver/config/GeoServerPropertyConfigurerTest.java (hunk 4 above)
Comment (dieterstueken) [reply src/main/src/test/java/org/geoserver/config/GeoServerPropertyConfigurerTest.java:41]: Took a while to realize it again. If you roll back to b14f54e1 testConfigFileCreated you can see the test to snap:

Without throwing a `FileNotFoundException` [GeoServerPropertyConfigurer.loadProperties()](https://github.com/geoserver/geoserver/blob/main/src/main/src/main/java/org/geoserver/config/GeoServerPropertyConfigurer.java#L104) won't ever enter the catch block, since an empty file was silently created, which did not hurt anybody else before.

The (otherwise failed) creation takes place within [GeoServerPropertyConfigurerTest.setUp()](https://github.com/geoserver/geoserver/pull/6314/files#diff-6e0586e9277979b3436a1ee5aa890935b97522db04e285e655279ce2d4ee236fR31) outside of testConfigFileCreated() and is hard to locate, since the exception is caught. Here is a stacktrace of the hidden exeption:

<pre>
java.io.FileNotFoundException: foo.properties
	at org.geoserver.config.SpringResourceAdaptor.getInputStream(SpringResourceAdaptor.java:62)
	at org.springframework.core.io.support.EncodedResource.getInputStream(EncodedResource.java:159)
	at org.springframework.core.io.support.PropertiesLoaderUtils.fillProperties(PropertiesLoaderUtils.java:99)
	at org.springframework.core.io.support.PropertiesLoaderSupport.loadProperties(PropertiesLoaderSupport.java:182)
--->	at org.geoserver.config.GeoServerPropertyConfigurer.loadProperties(GeoServerPropertyConfigurer.java:103)
	at org.springframework.core.io.support.PropertiesLoaderSupport.mergeProperties(PropertiesLoaderSupport.java:163)
	at org.springframework.context.support.PropertySourcesPlaceholderConfigurer.postProcessBeanFactory(PropertySourcesPlaceholderConfigurer.java:144)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:291)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:167)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:707)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:533)
	at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:202)
	at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:177)
	at org.springframework.context.support.ClassPathXmlApplicationContext.<init>(ClassPathXmlApplicationContext.java:163)
	at org.geoserver.config.GeoServerPropertyConfigurerTest.setUp(GeoServerPropertyConfigurerTest.java:31)
....
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
</pre>

So, after introducing the FNF Exception, creation of the missing property file works now and the unit test verifies it. I think this was hidden bug and went wrong before.

**Comment by aaime** [comment]
@jodygarnett are you happy to merge the request now? Dieter followed up with your question.

**Comment by dieterstueken** [comment]
yes, please!
no back port necessary for me.
